Source/Test/Rewriting/Passes/AssemblyRewriter.cs

```diff
@@ -468,124 +516,36 @@ protected static TypeDefinition Resolve(TypeReference type)
             return result;
         }
 
-        /// <summary>
-        /// Create a GenericInstanceType for the given generic type instantiated with the given generic arguments.
-        /// </summary>
-        /// <param name="module">The module we are operating on.</param>
-        /// <param name="genericType">The generic type to instantiate.</param>
-        /// <param name="genericArgs">The generic arguments needed to instantiate the generic type.</param>
-        /// <returns>The new generic instance type.</returns>
-        protected static GenericInstanceType ImportGenericTypeInstance(ModuleDefinition module, TypeReference genericType,
-            params TypeReference[] genericArgs)
+        public static TypeReference MakeGenericType(TypeReference self, params TypeReference[] arguments)
         {
-            TypeReference typeDef = genericType.Resolve();
-            if (!typeDef.HasGenericParameters)
+            if (self.GenericParameters.Count != arguments.Length)
             {
-                throw new InvalidOperationException(string.Format("Type {0} is not generic", genericType));
+                throw new ArgumentException();
             }
 
-            typeDef = module.ImportReference(typeDef);
-            var instance = new GenericInstanceType(typeDef);
-            for (int i = 0; i < typeDef.GenericParameters.Count; i++)
+            var instance = new GenericInstanceType(self);
+            foreach (var argument in arguments)
             {
-                var p = typeDef.GenericParameters[i];
-                if (p.Position < genericArgs.Length)
-                {
-                    GenericParameter parameter = new GenericParameter(p.Name, typeDef);
-                    instance.GenericParameters.Add(parameter);
-                    instance.GenericArguments.Add(genericArgs[p.Position]);
```

Comment:
I see, so this was not really required after all, and we could just add the specified input generic arguments to the new generic type to construct it ...